fix: Improve automatic login from refresh token

lfaucon opened the pull request:
* Create PublicRoute component
* Add tests for PublicRoute component
* Move usage of refresh token from Login to PublicRoute
* Reorganize a few imports

Closes #323

amatissart reviewed (commented):
Great. Except a comment about potential concurrent requests, it looks good to me and will make the UI much more consistent when a user is back after a previous session.

I think that there are still edge cases that are not covered, when the app triggers API requests soon after the access token expires. Refreshing the token early (before it has effectively expired) could help to mitigate the issue.

lfaucon commented on frontend/src/features/login/PublicRoute.tsx:
```diff
@@ -0,0 +1,29 @@
+const PublicRoute = (props: Props) => {
+  const login: LoginState = useAppSelector(selectLogin);
+  const dispatch = useAppDispatch();
+  const [shouldTryRefresh, setShouldTryRefresh] = useState(true);
```
No, I don't. I mindlessly copied it from the implementation inside the Login component. I see the difference, and it seems indeed not necessary here.

lfaucon commented on frontend/src/features/login/PublicRoute.tsx:
```diff
@@ -0,0 +1,29 @@
+    if (shouldTryRefresh && login.refresh_token && !isLoggedIn(login)) {
+      console.log('token invalid but refresh token present, trying to refresh');
+      setShouldTryRefresh(false);
+      dispatch(getTokenFromRefreshAsync(login.refresh_token));
```
I'll make this change! Although, I have removed the token refresh from the Login component to avoid this.

lfaucon commented:
Thanks for the review @amatissart. I addressed your comments with the last commit and chose to refresh the token if it expires in less than 10 minutes. Let me know if you think a shorter window would be better.

amatissart commented:
> Thanks for the review @amatissart. I addressed your comments with the last commit and chose to refresh the token if it expires in less than 10 minutes. Let me know if you think a shorter window would be better.

10 minutes is fine. We might even set a longer window, as the access token is currently valid for 10 hours.
